File: src/engine/combos.js
import { buildDeck } from "../cards.js";

/* ============================================================================
 * Combo metadata/helpers
 * - Centralized combo month maps from deck definition
 * - Card-tag and missing-month utility helpers
 * ========================================================================== */

const COMBO_KEYS = ["redRibbons", "blueRibbons", "plainRibbons", "fiveBirds"];

/* 1) Card/tag helpers */
function uniqueCards(cards = []) {
  const seen = new Set();
  return cards.filter((card) => {
    if (!card) return false;
    const key = card.id || `${card.month}:${card.category}:${card.name || ""}`;
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}

export function hasComboTag(card, tag) {
  return Array.isArray(card?.comboTags) && card.comboTags.includes(tag);
}

export function countComboTag(cards = [], tag) {
  return uniqueCards(cards).reduce((count, card) => (hasComboTag(card, tag) ? count + 1 : count), 0);
}

/* 2) Combo month maps built once at module load */
function buildComboMonthMap() {
  const map = Object.fromEntries(COMBO_KEYS.map((tag) => [tag, new Set()]));
  for (const card of buildDeck()) {
    const tags = Array.isArray(card.comboTags) ? card.comboTags : [];
    for (const tag of tags) {
      if (map[tag]) map[tag].add(card.month);
    }
  }
  return Object.fromEntries(
    COMBO_KEYS.map((tag) => [tag, Array.from(map[tag]).sort((a, b) => a - b)])
  );
}

const comboMonthMap = buildComboMonthMap();

export const COMBO_MONTHS = Object.freeze(comboMonthMap);

export const COMBO_MONTH_SETS = Object.freeze(
  Object.fromEntries(COMBO_KEYS.map((tag) => [tag, new Set(COMBO_MONTHS[tag])]))
);

/* 3) Public lookup helpers */
export function missingComboMonths(cards = [], tag) {
  const months = COMBO_MONTHS[tag] || [];
  const ownMonths = new Set(
    uniqueCards(cards)
      .filter((card) => hasComboTag(card, tag))
      .map((card) => card.month)
  );
  return months.filter((month) => !ownMonths.has(month));
}

File: experiments/k_hyperneat_matgo/py_eval/combos.py
from __future__ import annotations

from .cards import build_deck
from .models import Card


COMBO_KEYS = ("redRibbons", "blueRibbons", "plainRibbons", "fiveBirds")


def unique_cards(cards: list[Card] | None = None) -> list[Card]:
    seen: set[str] = set()
    out: list[Card] = []
    for card in cards or []:
        if not card:
            continue
        key = card.id or f"{card.month}:{card.category}:{card.name}"
        if key in seen:
            continue
        seen.add(key)
        out.append(card)
    return out


def has_combo_tag(card: Card | None, tag: str) -> bool:
    return bool(card and tag in card.combo_tags)


def count_combo_tag(cards: list[Card] | None, tag: str) -> int:
    return sum(1 for card in unique_cards(cards) if has_combo_tag(card, tag))


def _build_combo_month_map() -> dict[str, tuple[int, ...]]:
    raw: dict[str, set[int]] = {tag: set() for tag in COMBO_KEYS}
    for card in build_deck():
        for tag in card.combo_tags:
            if tag in raw:
                raw[tag].add(card.month)
    return {tag: tuple(sorted(raw[tag])) for tag in COMBO_KEYS}


COMBO_MONTHS = _build_combo_month_map()
COMBO_MONTH_SETS = {tag: set(months) for tag, months in COMBO_MONTHS.items()}


def missing_combo_months(cards: list[Card] | None, tag: str) -> list[int]:
    months = COMBO_MONTHS.get(tag, ())
    own_months = {card.month for card in unique_cards(cards) if has_combo_tag(card, tag)}
    return [month for month in months if month not in own_months]
